### PR TITLE
API Error Changes: Updates to Error Object

### PR DESCRIPTION
**Pull request by pkanher617**
Please review PR #586 before reviewing this.

This set of changes includes:
- Fixed stack variable and added tests to ensure stack trace was working
- Made errorMessage optional

### CODE REVIEW

File: lib/msal-core/src/error/AuthError.ts [hunk 1]
```diff
@@ -36,12 +36,11 @@ export class AuthError extends Error {
     errorCode: string;
     errorMessage: string;
 
-    constructor(errorCode: string, errorMessage: string) {
+    constructor(errorCode: string, errorMessage?: string) {
```
Comment (DarylThayil) [inline lib/msal-core/src/error/AuthError.ts]: just doublechecking that we don't wanna set a default here? does anything assume error message is required in this class / subclass

File: lib/msal-core/tests/error/error.spec.ts [hunk 2]
```diff
@@ -8,74 +8,105 @@ describe("Error", () => {
     const TEST_ERROR_CODE: string = "test";
     const TEST_ERROR_MSG: string = "This is a test error";
 
+    let throwFunc = (err: AuthError): void => {
```
Comment (DarylThayil) [inline lib/msal-core/tests/error/error.spec.ts]: const

File: lib/msal-core/tests/error/error.spec.ts [hunk 3]
```diff
@@ -8,74 +8,105 @@ describe("Error", () => {
...
     describe("ClientAuthError", () => {
         it("can be created", () => {
             let clientAuthError = new ClientAuthError(TEST_ERROR_CODE, TEST_ERROR_MSG);
+            let err: AuthError;
+            try {
+                throwFunc(clientAuthError);
+            } catch(error) {
+                err = error;
+            }
+            console.log("Stack trace: " + err.stack);
```
Comment (DarylThayil) [inline lib/msal-core/tests/error/error.spec.ts]: do we want to leave the logs in the tests?

File: lib/msal-core/tests/error/error.spec.ts (hunk 2 above)
Comment (DarylThayil) [reply lib/msal-core/tests/error/error.spec.ts]: what does this do? cant you just throw an AuthError below if all you are doing is try catching it

File: lib/msal-core/src/UserAgentApplication.ts [hunk 4]
```diff
@@ -364,7 +364,12 @@ export class UserAgentApplication {
     }
 
     // Validate and filter scopes (the validate function will throw if validation fails)
-    this.validateInputScope(scopes, false);
+    try {
```
Comment (manoj-rath) [inline lib/msal-core/src/UserAgentApplication.ts]: rethrowing adds unnecessary complexity in the code. We shouldn't use this, and there are multiple locations for this.

File: lib/msal-core/src/UserAgentApplication.ts [hunk 5]
```diff
@@ -714,7 +731,7 @@ export class UserAgentApplication {
 
       // If already in progress, throw an error and reject the request
       if (this._acquireTokenInProgress) {
-        return reject(ClientAuthError.createAcquireTokenInProgressError());
+        reject(ClientAuthError.createAcquireTokenInProgressError());
```
Comment (manoj-rath) [inline lib/msal-core/src/UserAgentApplication.ts]: is this intentional to remove the return?

File: lib/msal-core/src/UserAgentApplication.ts [hunk 6]
```diff
@@ -952,7 +973,7 @@ export class UserAgentApplication {
       //if user is not currently logged in and no login_hint/sid is passed as an extraQueryParamater
       if (!userObject && Utils.checkSSO(extraQueryParameters) && Utils.isEmpty(adalIdToken) ) {
         this._logger.info("User login is required");
-        return reject(ClientAuthError.createUserLoginRequiredError());
+        reject(ClientAuthError.createUserLoginRequiredError());
```
Comment (manoj-rath) [inline lib/msal-core/src/UserAgentApplication.ts]: same as above

**Pull request merged**